File: ios/Sources/Mia21/Services/SpaceService.swift
//
//  SpaceService.swift
//  Mia21
//
//  Created on November 4, 2025.
//  Copyright © 2025 Mia21. All rights reserved.
//
//  Description:
//  Service layer for space management operations.
//  Handles listing and retrieving workspace information.
//

import Foundation

// MARK: - Space Service Protocol

protocol SpaceServiceProtocol {
  func listSpaces() async throws -> [Space]
  func listBots() async throws -> [Bot]
  func listSpaceConversations(spaceId: String, options: SpaceConversationsOptions) async throws -> SpaceConversationsResponse
}

// MARK: - Space Service Implementation

final class SpaceService: SpaceServiceProtocol {

  // MARK: - Properties

  private let apiClient: APIClientProtocol

  // MARK: - Initialization

  init(apiClient: APIClientProtocol) {
    self.apiClient = apiClient
  }

  // MARK: - Public Methods

  func listSpaces() async throws -> [Space] {
    let endpoint = APIEndpoint(
      path: "/spaces",
      method: .get,
      body: nil
    )

    let spaces: [Space] = try await apiClient.performRequest(endpoint)
    return spaces
  }
  
  func listBots() async throws -> [Bot] {
    let endpoint = APIEndpoint(
      path: "/bots",
      method: .get,
      body: nil
    )

    // Response has format: {"bots": [...], "count": X}
    let response: BotsResponse = try await apiClient.performRequest(endpoint)
    return response.bots
  }
  
  func listSpaceConversations(spaceId: String, options: SpaceConversationsOptions) async throws -> SpaceConversationsResponse {
    logInfo("Listing conversations for space: \(spaceId)")
    
    // Build query parameters
    var queryParams: [String] = []
    
    if let userId = options.userId {
      queryParams.append("user_id=\(userId)")
    }
    if let botId = options.botId {
      queryParams.append("bot_id=\(botId)")
    }
    if let status = options.status {
      queryParams.append("status=\(status.rawValue)")
    }
    queryParams.append("limit=\(options.limit)")
    queryParams.append("offset=\(options.offset)")
    
    let queryString = queryParams.isEmpty ? "" : "?\(queryParams.joined(separator: "&"))"
    let path = "/spaces/\(spaceId)/conversations\(queryString)"
    
    let endpoint = APIEndpoint(
      path: path,
      method: .get,
      body: nil
    )
    
    let response: SpaceConversationsResponse = try await apiClient.performRequest(endpoint)
    logDebug("Found \(response.totalCount) conversations in space \(spaceId)")
    return response
  }
}

File: android/lib/src/main/kotlin/com/mia21/services/SpaceService.kt
/**
 * SpaceService.kt
 * Mia21
 *
 * Created on November 25, 2025.
 * Copyright © 2025 Mia21. All rights reserved.
 *
 * Description:
 * Service layer for space and bot management operations.
 */

package com.mia21.services

import com.mia21.models.Bot
import com.mia21.models.BotsResponse
import com.mia21.models.Space
import com.mia21.network.APIClient
import com.mia21.network.APIEndpoint
import com.mia21.network.HTTPMethod
import kotlinx.serialization.decodeFromString

/**
 * Service for managing spaces and bots
 */
class SpaceService(private val apiClient: APIClient) {
    
    /**
     * List all available spaces
     */
    suspend fun listSpaces(): List<Space> {
        val endpoint = APIEndpoint(
            path = "/spaces",
            method = HTTPMethod.GET
        )
        
        val jsonResponse = apiClient.performRequest(endpoint, String::class.java)
        return try {
            apiClient.json.decodeFromString<List<Space>>(jsonResponse)
        } catch (e: Exception) {
            emptyList()
        }
    }
    
    /**
     * List all bots for the current customer
     */
    suspend fun listBots(): List<Bot> {
        val endpoint = APIEndpoint(
            path = "/bots",
            method = HTTPMethod.GET
        )
        
        val jsonResponse = apiClient.performRequest(endpoint, String::class.java)
        val response = apiClient.json.decodeFromString<BotsResponse>(jsonResponse)
        return response.bots
    }
    
    /**
     * List conversations within a space.
     * Useful for admin dashboards, analytics, and bulk operations.
     */
    suspend fun listSpaceConversations(
        spaceId: String,
        options: SpaceConversationsOptions
    ): SpaceConversationsResponse {
        Logger.debug("Listing conversations for space: $spaceId")
        
        // Build query parameters
        val queryParams = mutableListOf<String>()
        
        options.userId?.let { queryParams.add("user_id=$it") }
        options.botId?.let { queryParams.add("bot_id=$it") }
        options.status?.let { queryParams.add("status=${it.value}") }
        queryParams.add("limit=${options.limit.coerceIn(1, 500)}")
        queryParams.add("offset=${maxOf(options.offset, 0)}")
        
        val queryString = if (queryParams.isNotEmpty()) "?${queryParams.joinToString("&")}" else ""
        val path = "/spaces/$spaceId/conversations$queryString"
        
        val endpoint = APIEndpoint(
            path = path,
            method = HTTPMethod.GET
        )
        
        val jsonResponse = apiClient.performRequest(endpoint, String::class.java)
        val response = apiClient.json.decodeFromString<SpaceConversationsResponse>(jsonResponse)
        Logger.debug("Found ${response.totalCount} conversations in space $spaceId")
        return response
    }
}

